### ISSUE TITLE
When running a release, the object type is modified

### ISSUE THREAD

**Issue by 415593725** [flutter/flutter#57472]
The debug release type is HttpResult,Why does the type change when the release runs?

**release mode log**
```bash
gMacBook-Pro:xxx xxxx$ flutter run -t lib/main_test.dart --release --flavor _test
Launching lib/main_test.dart on Redmi 7 in release mode...
Running Gradle task 'assemble_testRelease'...                           
Running Gradle task 'assemble_testRelease'... Done                 91.6s
✓ Built build/app/outputs/apk/_test/release/app-_test-release.apk (29.4MB).
Installing build/app/outputs/apk/app.apk...                         6.2s

Flutter run key commands.
h Repeat this help message.
c Clear the screen
q Quit (terminate the application on the device).
I/flutter (28711): runMode:false     result_runType: Mr

```

**debug mode log**
`I/flutter (29228): runMode:true     result_runType: HttpResult`

```dart
// json resolve
class JsonConvert {
  JsonConvert._();

  static T fromJsonAsT<T>(HttpResult result) {
     //The debug release type is HttpResult,Why does the type change when the release runs?
      print('runMode:$kDebugMode     result_runType: ${result.runtimeType}');

      switch (T.toString()) {
      case "LoginResp":
        return LoginResp.fromJson(result.data) as T;
        //............
        //............
     }
  }
}
```

```dart
//dio request
var response = await _dio.request(path,
        data: data, queryParameters: query, options: option..method = method, cancelToken: cancelToken);
    var httpResult = HttpResult.fromJson(response.data);
    return JsonConvert.fromJsonAsT<T>(httpResult);
```

```dart
//HttpResult
@JsonSerializable()
class HttpResult {
  @JsonKey(name: "code")
  int code;
  @JsonKey(name: "desc")//描述
  String desc;
  @JsonKey(name: "info") //信息
  String msg;
  @JsonKey(name: "data")//数据层
  dynamic data;

  bool get isSuccess => 200 == code;

  HttpResult({this.code, this.desc, this.msg, this.data});

  factory HttpResult.fromJson(Map<String, dynamic> json) => _$HttpResultFromJson(json);

  Map<String, dynamic> toJson() => _$HttpResultToJson(this);
}
```

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.17.1, on Mac OS X 10.15.4 19E266, locale zh-Hans-CN)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
[✓] Android Studio (version 3.6)
[!] IntelliJ IDEA Ultimate Edition (version 2020.1)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.45.0)
[✓] Connected device (1 available)

! Doctor found issues in 1 category.
```

**Comment by iapicca** [comment]
Hi @415593725 
I'm testing the code below with the latest master

<details>
  <summary>code sample</summary>

```dart
import 'package:flutter/material.dart';

class MysteryObject<T> {  
  String call() {
   final String _type = '$T';
   print(_type);
   return _type;
  }
}

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) => MaterialApp(
    home: MyHomePage(),
  );
}

class MyHomePage extends StatelessWidget {

  final MysteryObject _object = MysteryObject<String>();

  @override
  Widget build(BuildContext context) => Material(
    child: Center(
      child: Text(_object())
    ),
  );
}
```

</details>

<details>
  <summary>logs</summary>

```bash
 nevercode@Nevercodes-Air  ~/Desktop/projects/master  flutter_master run                                                                                   ✔  1403  15:27:36
Launching lib/main.dart on Pixel 3a in debug mode...
Running Gradle task 'assembleDebug'...
Running Gradle task 'assembleDebug'... Done                        13.0s
✓ Built build/app/outputs/flutter-apk/app-debug.apk.
Waiting for Pixel 3a to report its views...                         18ms
Syncing files to device Pixel 3a...                                266ms

Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h Repeat this help message.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).
An Observatory debugger and profiler on Pixel 3a is available at: http://127.0.0.1:59535/cm4HFd8cKFQ=/
I/flutter (26864): String
```

```bash
 nevercode@Nevercodes-Air  ~/Desktop/projects/master  flutter_master run --release                                                                         ✔  1404  15:29:00
Launching lib/main.dart on Pixel 3a in release mode...
Calling mockable JAR artifact transform to create file: /Users/nevercode/.gradle/caches/transforms-2/files-2.1/83d3107cfd2601930fdcaea8c808ac7d/android.jar with input /Users/nevercode/Library/Android/sdk/platforms/android-28/android.jar
Running Gradle task 'assembleRelease'...
Running Gradle task 'assembleRelease'... Done                      90.9s
✓ Built build/app/outputs/flutter-apk/app-release.apk (6.2MB).
Installing build/app/outputs/flutter-apk/app.apk...                 3.2s

Flutter run key commands.
h Repeat this help message.
c Clear the screen
q Quit (terminate the application on the device).
I/flutter (27139): String
```


</details>


<details>
  <summary>doctor</summary>

```console
[✓] Flutter (Channel master, 1.19.0-2.0.pre.90, on Mac OS X 10.15.4 19E287, locale en-EE)
    • Flutter version 1.19.0-2.0.pre.90 at /Users/nevercode/development/flutter_master
    • Framework revision f64f6e2b6b (3 hours ago), 2020-05-18 05:42:01 -0400
    • Engine revision 2bd71fb60a
    • Dart version 2.9.0 (build 2.9.0-9.0.dev b560016446)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/nevercode/Library/Android/sdk
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    • CocoaPods version 1.9.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.10.2

[✓] Connected device (3 available)
    • Pixel 3a   • 965AY0WP5C • android-arm64  • Android 10 (API 29)
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 81.0.4044.138

• No issues found!
```

</details>

does this code represent the issue you are describing,
if it doesn't please provide a minimal reproducible code sample
(without 3rd party packages or plugins)

Thank you

**Comment by 415593725** [comment]
Found the reason

when “gradle.properties” adds to “extra-gen-snapshot-options=--obfuscate”，
there will be that problem
